File: codes/2024/25-Excellence-BNU-99-21-43 (C++)/workspace/source/backend/controller/LifeTimeIndex.cpp
#include "LifeTimeIndex.h"
#include "LifeTimeController.h"

namespace Backend
{
    LifeTimeIndex::Type LifeTimeIndex::getType() const
    {
        return type;
    }

    bool LifeTimeIndex::isIn() const
    {
        return type == LifeTimeIndex::Type::IN;
    }

    bool LifeTimeIndex::isOut() const
    {
        return type == LifeTimeIndex::Type::OUT;
    }

    LifeTimeController *LifeTimeIndex::getLifeTimeController() const
    {
        return ltc;
    }

    int LifeTimeIndex::getInstID() const
    {
        return getLifeTimeController()->getInstID(srcInst);
    }

    AsmInstBasic *LifeTimeIndex::getSrcInst()
    {
        return srcInst;
    }

    LifeTimeIndex *LifeTimeIndex::getInstIn(LifeTimeController *ltcontroller, AsmInstBasic *srcinst)
    {
        return new LifeTimeIndex(ltcontroller, srcinst, LifeTimeIndex::Type::IN);
    }

    LifeTimeIndex *LifeTimeIndex::getInstOut(LifeTimeController *ltcontroller, AsmInstBasic *srcinst)
    {
        return new LifeTimeIndex(ltcontroller, srcinst, LifeTimeIndex::Type::OUT);
    }

    bool LifeTimeIndex::operator<(const LifeTimeIndex &other) const
    {
        if (getInstID() != other.getInstID())
            return getInstID() < other.getInstID();
        return type == Type::IN && other.type == Type::OUT;
    }

    bool LifeTimeIndex::operator==(const LifeTimeIndex &other) const
    {
        return getInstID() == other.getInstID() && type == other.type;
    }

    bool LifeTimeIndex::operator!=(const LifeTimeIndex &other) const
    {
        return !(*this == other);
    }

    bool LifeTimeIndex::operator<=(const LifeTimeIndex &other) const
    {
        return *this < other || *this == other;
    }

    std::string LifeTimeIndex::toString() const
    {
        return std::to_string(getInstID()) + "." + (type == LifeTimeIndex::Type::IN ? "IN" : "OUT");
    }
}

File: codes/2023/2-First-BIT (Java)/src/main/java/cn/edu/bit/newnewcc/backend/asm/controller/LifeTimeIndex.java
package cn.edu.bit.newnewcc.backend.asm.controller;

import cn.edu.bit.newnewcc.backend.asm.instruction.AsmInstruction;

public class LifeTimeIndex implements Comparable<LifeTimeIndex> {
    public enum Type {
        IN, OUT
    }

    private final LifeTimeController lifeTimeController;
    private final AsmInstruction sourceInst;
    private final Type type;

    public Type getType() {
        return type;
    }

    public boolean isIn() {
        return type == Type.IN;
    }

    public boolean isOut() {
        return type == Type.OUT;
    }

    private LifeTimeIndex(LifeTimeController lifeTimeController, AsmInstruction sourceInst, Type type) {
        this.lifeTimeController = lifeTimeController;
        this.sourceInst = sourceInst;
        this.type = type;
    }

    public int getInstID() {
        return lifeTimeController.getInstID(sourceInst);
    }

    public AsmInstruction getSourceInst() {
        return sourceInst;
    }

    static LifeTimeIndex getInstIn(LifeTimeController lifeTimeController, AsmInstruction sourceInst) {
        return new LifeTimeIndex(lifeTimeController, sourceInst, Type.IN);
    }

    static LifeTimeIndex getInstOut(LifeTimeController lifeTimeController, AsmInstruction sourceInst) {
        return new LifeTimeIndex(lifeTimeController, sourceInst, Type.OUT);
    }

    @Override
    public int compareTo(LifeTimeIndex o) {
        if (getInstID() != o.getInstID()) {
            return getInstID() - o.getInstID();
        }
        if (type != o.type) {
            return type == Type.IN ? -1 : 1;
        }
        return 0;
    }

    @Override
    public String toString() {
        return getInstID() + "." + getType();
    }
}
